src/metrics-gatherer.ts

```diff
@@ -43,29 +79,16 @@ export class MetricsGatherer {
 		this.metrics = {
 			gauge: {},
 			counter: {},
-			summary: {},
 			histogram: {},
+			summary: {},
 		};
-		this.customParams = {};
-		this.descriptions = {};
-		this.existMap = {};
-		this.kinds = {};
-	}
-
-	// fetch the description for a metric
-	public describe(name: string, text: string, custom: CustomParams = {}) {
-		if (this.descriptions[name]) {
-			throw new MetricsGathererError(
-				`tried to describe metric "${name}" twice`,
-			);
-		}
-		this.descriptions[name] = text;
-		this.customParams[name] = custom;
+		this.meta = {};
+		this.describe = new MetricsDescriber(this);
```

Comment:
This makes the two classes heavily interlinked in a way that it really throws up the question of why they're two separate classes, since they have a 1-1 relationship and neither has any meaning in isolation.  I think it probably makes sense to just have `describe` accept `Kind` as a parameter to avoid it

Reply:
@Page- yeah, I initially had `kind` as a parameter, but it's more clumsy to type and doesn't throw a compile-time error if it's wrong. I can't think of another way to get compile-time checking other than requiring the user to import an enum, which is cumbersome.

Reply:
Importing an enum is fine imo, besides that if it's a string based enum then they should also be able to use the strings directly and it will be accepted (assuming the string is valid)

Reply:
1
```
import { metrics } from '@balena/node-metrics-gatherer'
metrics.describe.counter('builds_total', 'number of builds initiated');
```

2
```
import { metrics, MetricsKind } from '@balena/node-metrics-gatherer'
metrics.describe(MetricsKind.COUNTER, 'builds_total', 'number of builds initiated');
```

Do you have strong feelings against 2? This is the alternative to 1, which is what the PR currently uses.

Reply:
The second can also be written as
```ts
import { metrics } from '@balena/node-metrics-gatherer'
metrics.describe('counter', 'builds_total', 'number of builds initiated');
```
btw. Since `MetricsKind.COUNTER === 'counter'` and it should still pass the type checks.

I think it'd also be ok to do:
```ts
this.describe = (
	kind: Kind,
	name: string,
	help: string,
	customParams: CustomParams = {}
) => {
	if (this.meta[name]) {
		throw new MetricsGathererError(
			`tried to describe metric "${name}" twice`,
		);
	} else {
		this.meta[name] = {
			kind,
			help,
			customParams,
		};
	}
}
for (const kind of ['gauge', 'counter', 'histogram', 'summary']) {
	this.describe[kind] = _.partial(this.describe, kind)
}
```
which preserves both interfaces and avoids the weird 1-1 class linking, but wouldn't make use of a prototype for describe

Reply:
This seems doable in JS but typescript is throwing all kinds of objections. `describe` needs to be in the prototype, so I added:

```
type DescribeFn = (
	kind: Kind,
	name: string,
	help: string,
	customParams: CustomParams,
) => void;

export class MetricsGatherer {
	public metrics: MetricsMap;
	public meta: MetricsMetaMap;
	public describe: DescribeFn;
...
```

But then we can't `this.describe[kind]` because `DescribeFn` has no index signature.

If we then instead do:

```
interface DescribeFn {
	(
		kind: MetricKind,
		name: string,
		help: string,
		customParams: CustomParams,
	): void;
	[kind: string]: (
		name: string,
		help: string,
		customParams: CustomParams,
	) => void;
}

export class MetricsGatherer {
	public metrics: MetricsMap;
	public meta: MetricsMetaMap;
	public describe: DescribeFn;

	constructor() {
		this.initState();
		this.setupDescribe();
	}

	private setupDescribe() {
		this.describe = describe = (
			kind: MetricKind,
			name: string,
			help: string,
			customParams: CustomParams = {},
		) => {
			if (this.meta[name]) {
				throw new MetricsGathererError(
					`tried to describe metric "${name}" twice`,
				);
			} else {
				this.meta[name] = {
					kind,
					help,
					customParams,
				};
			}
		};
		for (const kind of ['gauge', 'counter', 'histogram', 'summary']) {
			this.describe[kind] = _.partial(this.describe, kind);
		}
	}
```

we get: 

```
Type '(kind: MetricKind, name: string, help: string, customParams?: CustomParams) => void' is not assignable to type 'DescribeFn'.
  Index signature is missing in type '(kind: MetricKind, name: string, help: string, customParams?: CustomParams) => void'.
```

This is kind of doing my head in haha

Reply:
if you use
```ts
or (const kind of ['gauge', 'counter', 'histogram', 'summary'] as const) {
```
it should avoid that index signature because the type of `kind` should change from `string` to `'gauge' | 'counter' | 'histogram' | 'summary'` which means it'll match the specific types.  Also if you just use `public describe` and allow for type inference it might "just work" because typescript is pretty good at inferring things (and generally the more you can let it infer the easier life will be for you)

Reply:
Ok, I just pushed up a much cleaner implementation of this. I have no idea why I was trying to put the describe function partial on an instance of a describe function itself. Makes no sense. Now we just have:

```
interface Describer {
	[kind: string]: (
		name: string,
		help: string,
		customParams?: CustomParams,
	) => void;
}
```

and then

```
                this.describe = {};
		for (const kind of ['gauge', 'counter', 'histogram', 'summary'] as const) {
			this.describe[kind] = ( ... function body ...
```

Reply:
Well, @dt-rush typing this as `[kind: string]: (...) => void;` removed any sort of static type guarantees because I can now do `metrics.describe.ampelia(...)` and it'll compile fine but will silently fail to register the metric at runtime. @Page- is right that there should be an enum like `gauge | counter | ...` so that the compiler can reason about the possible keys.